include/deal.II/lac/la_parallel_vector.templates.h

```diff
@@ -629,79 +631,43 @@ namespace LinearAlgebra
+    operator=(const Vector<Number, MemorySpaceType2> &c)
     {
-      Assert(c.partitioner.get() != nullptr, ExcNotInitialized());
-
       // we update ghost values whenever one of the input or output vector
       // already held ghost values or when we import data from a vector with
       // the same local range but different ghost layout
-      bool must_update_ghost_values = c.vector_is_ghosted;
+      assignment(c, c.vector_is_ghosted);
 
-      // check whether the two vectors use the same parallel partitioner. if
-      // not, check if all local ranges are the same (that way, we can
-      // exchange data between different parallel layouts). One variant which
-      // is included here and necessary for compatibility with the other
-      // distributed vector classes (Trilinos, PETSc) is the case when vector
-      // c does not have any ghosts (constructed without ghost elements given)
-      // but the current vector does: In that case, we need to exchange data
-      // also when none of the two vector had updated its ghost values before.
-      if (partitioner.get() == nullptr)
-        reinit(c, true);
-      else if (partitioner.get() != c.partitioner.get())
-        {
-          // local ranges are also the same if both partitioners are empty
-          // (even if they happen to define the empty range as [0,0) or [c,c)
-          // for some c!=0 in a different way).
-          int local_ranges_are_identical =
-            (partitioner->local_range() == c.partitioner->local_range() ||
-             (partitioner->local_range().second ==
-                partitioner->local_range().first &&
-              c.partitioner->local_range().second ==
-                c.partitioner->local_range().first));
-          if ((c.partitioner->n_mpi_processes() > 1 &&
-               Utilities::MPI::min(local_ranges_are_identical,
-                                   c.partitioner->get_mpi_communicator()) ==
-                 0) ||
-              !local_ranges_are_identical)
-            reinit(c, true);
-          else
-            must_update_ghost_values |= vector_is_ghosted;
+      return *this;
+    }
 
-          must_update_ghost_values |=
-            (c.partitioner->ghost_indices_initialized() == false &&
-             partitioner->ghost_indices_initialized() == true);
-        }
-      else
-        must_update_ghost_values |= vector_is_ghosted;
 
-      thread_loop_partitioner = c.thread_loop_partitioner;
 
-      const size_type this_size = partitioner->local_size();
-      if (this_size > 0)
-        {
-          dealii::internal::VectorOperations::
-            functions<Number, Number2, MemorySpaceType>::copy(
-              thread_loop_partitioner, this_size, c.data, data);
-        }
+    template <typename Number, typename MemorySpaceType>
+    template <typename Number2>
+    inline Vector<Number, MemorySpaceType> &
+    Vector<Number, MemorySpaceType>::
+    operator=(const Vector<Number2, MemorySpaceType> &c)
```

Comment:
Can't you just have a
```
operator=(const Vector<Number2, MemorySpaceType2> &c)
```
signature for all the combinations and `static_assert` that the `MemorySpace` is `Host` if we don't compile with `CUDA` support?

Reply:
No that doesn't compile. I need at least two functions `operator=(const Vector<Number, MemorySpaceType> &c)` (otherwise I can a bunch of error that the operator is deleted) and `operator=(const Vector<Number2, MemorySpaceType2> &c)`. I think it's clearer this way too since I do not allow `Number != Number2` and `MemorySpaceType != MemorySpaceType2` at the same time.